# Использование NumPy

Импортируем необходимые библиотеки

In [54]:
import numpy as np
import time
import random
import pandas as pd

## Генерация массивов

Генерируем линейный массив чисел с плавающей точкой размерностью (1,20)

In [68]:
np.random.rand(20)

array([8.69196143e-01, 8.21762997e-01, 7.51224166e-01, 7.60661858e-01,
       1.23927295e-01, 8.40858222e-01, 2.36344337e-01, 5.15778687e-01,
       6.06518529e-01, 3.08103610e-01, 5.40501317e-01, 9.16888692e-01,
       6.06433209e-01, 3.44987453e-01, 3.76282334e-02, 7.10313594e-04,
       8.04805533e-01, 9.84567466e-01, 2.11747355e-01, 2.50465634e-01])

Далее создаем функцию генерации одномерных массивов и оцениваем время, затраченного на генерацию

In [34]:
def generate_array_numpy():
    count_list = [100,1000,100000,1000000,100000000]
    for count in count_list:
        start_time = time.time()
        np.random.rand(count)
        
        time_now = (time.time() - start_time)
        time_now = round(time_now, 3)
        print(f"{count} элементов генерируется {time_now} секунд")

In [35]:
generate_array_numpy()

100 элементов генерируется 0.0 секунд
1000 элементов генерируется 0.0 секунд
100000 элементов генерируется 0.001 секунд
1000000 элементов генерируется 0.016 секунд
100000000 элементов генерируется 1.487 секунд


Генерируем двумерный массив чисел с плавающей точкой размерностью (2,10)

In [69]:
np.random.rand(2,10)

array([[0.32522623, 0.55025363, 0.00859578, 0.14154292, 0.55707401,
        0.78456337, 0.00908645, 0.43703931, 0.65410934, 0.27324417],
       [0.1627704 , 0.15167047, 0.62567318, 0.91377996, 0.18434153,
        0.17548984, 0.85185595, 0.30061704, 0.82020364, 0.00168318]])

Функция генерации двумерных массивов с оценкой затраченного времени

In [36]:
def generate_matrix_numpy():
    count_list = [100,1000,100000,1000000,100000000]
    for count in count_list:
        start_time = time.time()
        np.random.rand(2, count)
        
        time_now = (time.time() - start_time)
        time_now = round(time_now, 3)
        print(f"{count} элементов генерируется {time_now} секунд")

In [37]:
generate_matrix_numpy()

100 элементов генерируется 0.0 секунд
1000 элементов генерируется 0.0 секунд
100000 элементов генерируется 0.01 секунд
1000000 элементов генерируется 0.025 секунд
100000000 элементов генерируется 2.825 секунд


Затем попробуем сгенерировать массивы с помощью стандартной библиотеки pyhon и сравнить время, которое затрачивается на генерацию массивов одинаковой размерности.

In [47]:
array_list = []
for i in range(20):
    array_list.append(random.random())
array_list

[0.05092981525271789,
 0.22844718374535766,
 0.24840237159268008,
 0.8376562099969839,
 0.12354418750376228,
 0.8426551935598487,
 0.1081439099289161,
 0.7642152002169148,
 0.2597372524555416,
 0.4797067292791384,
 0.39200827964338436,
 0.7161300984513246,
 0.677931090814924,
 0.9920918603845617,
 0.45369826456363505,
 0.7945888659429452,
 0.8714261270834369,
 0.3065170817080841,
 0.05445781392036231,
 0.8783885693418022]

In [49]:
def generate_array_python():
    count_list = [100,1000,100000,1000000,100000000]
    for count in count_list:
        start_time = time.time()
        
        array_list = []
        for iter in range(count):
            array_list.append(random.random())
        
        time_now = (time.time() - start_time)
        time_now = round(time_now, 3)
        print(f"{count} элементов генерируется {time_now} секунд")

In [50]:
generate_matrix_python()

100 элементов генерируется 0.0 секунд
1000 элементов генерируется 0.0 секунд
100000 элементов генерируется 0.025 секунд
1000000 элементов генерируется 0.268 секунд
100000000 элементов генерируется 25.75 секунд


**На генерацию массивов с помощью стандартной библиотеки python уходит намного больше времени, чем используя numpy. Так например, чтобы создать одномерный массив размерностью (1, 1000000), у python уходит 25.75 секунд, а у numpy всего 2.825**

## Арифметические операции над элементами двумерного массива

Создадим 2 матрицы размера (2,5)

In [70]:
rand_matrix1 = np.random.rand(2,5)
rand_matrix2 = np.random.rand(2,5)

Для удобства восприятия выводим массивы в качестве датафрейма

In [71]:
pd.DataFrame(rand_matrix1)

0         1         2         3         4
0  0.286024  0.854715  0.538198  0.723858  0.044789
1  0.711682  0.374337  0.083428  0.226819  0.438445

In [72]:
pd.DataFrame(rand_matrix2)

0         1         2         3         4
0  0.674485  0.198501  0.560355  0.591097  0.748729
1  0.323236  0.592268  0.005874  0.198027  0.878524

* Умножение на число

In [73]:
number_multipl_matrix = 5 * rand_matrix1
pd.DataFrame(number_multipl_matrix)

0         1         2         3         4
0  1.430121  4.273576  2.690989  3.619288  0.223943
1  3.558408  1.871683  0.417138  1.134094  2.192224

* Сложение матриц

In [74]:
sum_matrix = rand_matrix1 + rand_matrix2
pd.DataFrame(sum_matrix)

0         1         2         3         4
0  0.960509  1.053216  1.098552  1.314955  0.793518
1  1.034917  0.966605  0.089302  0.424846  1.316969

* Разность матриц

In [75]:
sum_matrix = rand_matrix1 - rand_matrix2
pd.DataFrame(sum_matrix)

0         1         2         3         4
0 -0.388461  0.656214 -0.022157  0.132760 -0.703940
1  0.388446 -0.217932  0.077553  0.028792 -0.440079

* Транспонирование матрицы

In [77]:
trans_matrix = rand_matrix2.T
pd.DataFrame(trans_matrix)

0         1
0  0.674485  0.323236
1  0.198501  0.592268
2  0.560355  0.005874
3  0.591097  0.198027
4  0.748729  0.878524

* Произведение матриц

In [78]:
multipl_matrix = rand_matrix1.dot(trans_matrix)
pd.DataFrame(multipl_matrix)

0         1
0  1.125567  0.784527
1  1.063422  0.882339